distributed/versions.py

```diff
@@ -131,12 +140,33 @@ def error_message(scheduler, workers, client, client_name="client"):
         )
         if len(versions) <= 1:
             continue
-        rows = [
-            (node_name, node_packages[node_name][pkg]) for node_name in nodes.keys()
-        ]
-        errs.append("%s\n%s" % (pkg, asciitable(["", "version"], rows)))
+
+        client_version = None
+        scheduler_version = None
+        workers_version = set()
+        for node_name in nodes.keys():
+            if node_name == "client":
+                client_version = node_packages[node_name][pkg]
+            elif node_name == "scheduler":
+                scheduler_version = node_packages[node_name][pkg]
+            else:
+                workers_version.add(node_packages[node_name][pkg])
+
+        if len(workers_version) == 1:
+            workers_version = list(workers_version)[0]
+        elif len(workers_version) == 0:
+            workers_version = None
```

Comment:
I think with the rewrite here a lot of this looping and dict building isn't really necessary anymore. Since we treat `client` and `scheduler` specially, there's probably no need to create the `nodes` dict above, then create the `node_packages` dict, then extract all that info out again. I think the logic we want is (pseudocode):

```python
# Much of this preprocessing is done in the `node_packages` building loop at the top currently
# I think we can refactor this to not be so complicated
client = ... # ensure client is a dict of packages
scheduler = ... # ensure scheduler is a dict of packages
workers = ... # ensure workers is a mapping of id -> dict of packages
# ...
packages = get_set_of_all_packages_provided(client, scheduler, worker)

for package in sorted(packages):
    if versions_dont_all_match():
        errs.append(...)
        if pkg in notes_mismatch_package:
            notes.append(...)
```

Reply:
@jcrist I did some rework on this. I appreciate your feedback. 
Thanks,